Remove Schema and State from linked list

LJBabbage opened the pull request:
### What is the context of this PR?

This PR removes the Schema and State from the linked list. Instead the node holds the id of the block and the answers associated with that block. When there is a need for the State (validation, or rendering) it finds it in the parsed version of the schema and generates the state on the fly, with any previous answers. Json pickle now only saves the linked list

Validation is now in one place, as it used to be under validation (Mandatory) and state update (Type Checking). We no longer redirect on failed validation, instead just recalling the renderer, meaning we don't have to reload the state again.

I have renamed a number of classes and functions as well to make the code more readable and easy to follow

**What has been removed**
All repeating elements and their tests
Other option from checkboxes and radio button, due to the use of state in linked list

**Behaviour change**
Error messages do not appear on a get request
### How to review

Full system testing

Commit:
Remove Schema from State and remove repeating

benjefferies commented on app/main/views/questionnaire.py:
```diff
@@ -92,9 +92,12 @@ def do_post(collection_id, eq_id, form_type, period_id, location, questionnaire_
+    if is_valid:
+        metadata = get_metadata(current_user)
+        logger.info("Redirecting user to next location %s with tx_id=%s", next_location, metadata.tx_id)
+        return redirect('/questionnaire/' + eq_id + '/' + form_type + '/' + period_id + '/' + collection_id + '/' + next_location)
```
do_redirect could be made use of here

benjefferies commented on app/questionnaire/questionnaire_manager.py:
```diff
@@ -184,78 +155,69 @@ def get_first_location(self):
-                answers_dict[answer.id] = answer.value
-            elif answer.other and isinstance(answer.value, list):
-                answer_values = answer.value[:]
-                if 'other' in answer_values:
```
Are we losing the 'other' functionality @samiwel introduced here

LJBabbage replied:
I pulled it out as it was using state on the find answer, he is rewriting it now

LJBabbage replied:
Sam has now rewritten this

benjefferies commented on app/questionnaire/questionnaire_manager.py:
```diff
@@ -184,78 +155,69 @@ def get_first_location(self):
+                is_valid = self.schema_item.validate(self.state)
+                if not is_valid:
+                    location = node.item_id
+                    logger.error("Failed validation with current location %s", location)
```
What was the reason to up the level of logging for a validation error from debug to error? Doesn't seem necessary to me.

LJBabbage replied:
True, left over from me fixing it

benjefferies commented on app/questionnaire/questionnaire_manager.py:
```diff
@@ -328,62 +284,71 @@ def process_incoming_answers(self, location, post_data, replay=False):
+        node = self._current
+        state_items = []
+        if is_valid:
+            if location == 'summary':
```
Would it be cleaner to break this out into get_summary_rendering_context and get_questionnaire_rendering_context then push the conditional logic further up?

benjefferies commented on tests/app/questionnaire/test_questionnaire_manager.py:
```diff
@@ -29,110 +31,20 @@ def setUp(self):
         block1 = self.questionnaire.children[0].children[0]
-        self.assertIsNone(questionnaire_manager.get_state(block1.id))
-
-    def test_single_value_answer_with_other_returns_other(self):
```
Looks like this may have been merged out?

LJBabbage replied:
Good spot

LJBabbage replied:
Sam has added the extra tests as well

benjefferies commented on tests/app/questionnaire/test_questionnaire_manager_factory.py:
```diff
@@ -14,45 +12,24 @@ class TestQuestionnaireManagerFactory(SurveyRunnerTestCase):
-
-    def mock_state(self):
-        StateManager.has_state = MagicMock(return_value="True")
+    def mock_user_journey(self):
```
This appears not to be used

benjefferies commented on tests/app/questionnaire/test_questionnaire_manager_factory.py:
```diff
@@ -14,45 +12,24 @@ class TestQuestionnaireManagerFactory(SurveyRunnerTestCase):
+        UserJourneyManager.get_user_journey = self.get_user_journey
         QuestionnaireManagerFactory._get_schema = self.get_schema
 
     def test_get_instance(self):
```
Not sure how useful this test's being now

benjefferies commented on tests/app/templating/test_summary_template_preprocessor.py:
```diff
@@ -55,26 +58,20 @@ def test_build_view_data(self):
         node1.next = node2
 
-        summary_template_preprocessor = SummaryTemplatePreprocessor()
+        questionnaire_template_preprocessor = SummaryTemplatePreprocessor()
```
Variable has been renamed incorrectly also test class name appears to be wrong?

benjefferies commented on tests/app/templating/test_summary_template_preprocessor.py:
```diff
@@ -55,26 +58,20 @@ def test_build_view_data(self):
+        state_items = [state]
+        render_data = questionnaire_template_preprocessor.build_view_data(node1, self.questionnaire, state_items)
         self.assertIsNotNone(render_data)
-        blocks = render_data['content']
```
Should we still have these assertions?

benjefferies commented on tests/integration/star_wars/test_radio_boxes.py:
```diff
@@ -85,34 +81,3 @@ def test_radio_boxes_mandatory_empty(self):
         content = resp.get_data(True)
         self.assertRegexpMatches(content, 'Why doesn&#39;t Chewbacca receive a medal at the end of A New Hope?')
-
-    def test_radio_boxes_other_is_mandatory(self):
```
Looks like this has been merged out

LJBabbage replied:
As above

LJBabbage commented:
Definitely some loose ends that need tiding up, the focus wasn't on getting it perfect, just working for people to work of. So I guess a 2 pronged approach, what needs to be done now to get it in and what we have to do but later

ajmaddaford commented on app/questionnaire/questionnaire_manager.py:
```diff
@@ -328,62 +284,71 @@ def process_incoming_answers(self, location, post_data, replay=False):
+        self.state = None
+        self.schema_item = None
+        if self._schema.item_exists(node.item_id):
+            self.schema_item = self._schema.get_item_by_id(node.item_id)
```
I don't think we need to set schema_item on self. Once we've created the state, isn't it accessible via `state.schema_item`?

ajmaddaford commented on app/questionnaire/questionnaire_manager.py:
```diff
@@ -328,62 +284,71 @@ def process_incoming_answers(self, location, post_data, replay=False):
+
+        return preprocessor.build_view_data(node, self._schema, state_items)
+
+    def build_state(self, node, answers):
```
We should minimise what we pass to methods. This method only uses the node to get the item_id (more specifically a block_id?), so it would be better to pass just that.

LJBabbage replied:
Yep, I used to use node.answer in this method, I removed the need for that, so you are correct it only uses the id

ajmaddaford commented on app/questionnaire/user_journey_manager.py:
```diff
@@ -0,0 +1,85 @@
+import logging
```
This functionality should be implemented along the same lines as metadata in data_model/questionnaire_store.

LJBabbage replied:
Not tackled as this was only name changes to the class and will be refactored out later anyway, as dicussed

ajmaddaford commented on app/schema/answers/date_answer.py:
```diff
@@ -12,7 +12,8 @@ def __init__(self, answer_id=None):
         self.widget = DateWidget(self.id)
 
-    def get_typed_value(self, post_data):
+    def type_validation(self, post_data):
```
I don't think this method should have been renamed - it returns a typed value.

LJBabbage replied:
True, but its also doing validation, this refers back to the note I made in the code, that validation and type value casting is being done in the same place which isn't right

samiwel committed:
Fix radio and checkbox other input field on spike branch.

samiwel committed:
Add unit tests for radio/checkbox other field.

iwootten commented on app/questionnaire_state/state_answer.py:
```diff
@@ -22,6 +22,18 @@ def update_state(self, user_input):
+        # Get radio options from the schema.
+        for option in self.schema_item.options:
+            # If the radio answer has an Other option...
+            if option['label'] == 'Other' and 'other' in option and self.input not in [option['value'] for option in self.schema_item.options]:
```
This if statement is pretty difficult to read, due to it's length. Do you think you could break the conditions out onto multiple lines @samiwel?

samiwel replied:
@iwootten I'll take a look, thanks.

iwootten commented on app/schema/widgets/multiple_choice_widget.py:
```diff
@@ -26,3 +26,12 @@ def render(self, answer_state):
+        if len(post_vars) == 0:
+            return None
+
+        other_values = post_vars.getlist(self.name) if hasattr(post_vars, 'getlist') else post_vars.get(self.name)
```
Similarly @samiwel, these two if conditions would be better if they were over multiple lines.

samiwel committed:
Remove Schema from State and remove repeating

iwootten commented on app/schema/widgets/multiple_choice_widget.py:
```diff
@@ -28,10 +28,23 @@ def render(self, answer_state):
+            other_value = last_element
+
         logger.debug('Getting user input for "{}", value is "{}"'.format(self.name, other_value))
         return other_value
```
@samiwel You could reduce this down a bit to:

```
return posted_data[-1:][0] if has_multiple_values else None
```

and lose the other_value assignments.

Commit:
Update after code review

Commit:
change redirect

benjefferies committed:
Remove tests for errors not displayed on GET page

We no longer store errors in the state nor do we recalculate errors on a
given page when a page is requested. The errors will be recalculated
when the page is submitted. For this reason the tests are no longer
valid so we are removing them.
